agent/consul/state/catalog_test.go

```diff
@@ -401,6 +453,189 @@ func TestStateStore_EnsureRegistration_Restore(t *testing.T) {
 	}()
 }
 
+func deprecatedEnsureNodeWithoutIDCanRegister(t *testing.T, s *Store, nodeName string, txIdx uint64) {
+	// All the following is deprecated, and should be removed in future Consul versions
+	in := &structs.Node{
+		Node:    nodeName,
+		Address: "1.1.1.9",
+	}
+	if err := s.EnsureNode(txIdx, in); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	idx, out, err := s.GetNode(nodeName)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	if idx != txIdx {
+		t.Fatalf("index should be %q, was: %q", txIdx, idx)
+	}
+	if out.Node != nodeName {
+		t.Fatalf("unexpected result out = %q, nodeName supposed to be %s", out, nodeName)
+	}
+}
+
+func TestStateStore_EnsureNodeDeprecated(t *testing.T) {
+	s := testStateStore(t)
+
+	firstNodeName := "node-without-id"
+	deprecatedEnsureNodeWithoutIDCanRegister(t, s, firstNodeName, 1)
+
+	newNodeID := types.NodeID("00a916bc-a357-4a19-b886-59419fcee50c")
+	// With this request, we basically add a node ID to existing node
+	// and change its address
+	in := &structs.Node{
+		ID:      newNodeID,
+		Node:    firstNodeName,
+		Address: "1.1.7.8",
+	}
+	if err := s.EnsureNode(4, in); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	// Retrieve the node again
+	idx, out, err := s.GetNode(firstNodeName)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	// Node has updated information
+	if idx != 4 || out.Node != firstNodeName || out.ID != newNodeID || out.Address != "1.1.7.8" {
+		t.Fatalf("[DEPRECATED] bad node returned: %#v", out)
+	}
+	if out.CreateIndex != 1 || out.ModifyIndex != 4 {
+		t.Fatalf("[DEPRECATED] bad CreateIndex/ModifyIndex returned: %#v", out)
+	}
+
+	// Now, lets update IP Address without providing any ID
+	// Only name of node will be used to match
+	in = &structs.Node{
+		Node:    firstNodeName,
+		Address: "1.1.7.10",
+	}
+	if err := s.EnsureNode(7, in); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	// Retrieve the node again
+	idx, out, err = s.GetNode(firstNodeName)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	// Node has updated information, its ID has been removed (deprecated, but working)
+	if idx != 7 || out.Node != firstNodeName || out.ID != "" || out.Address != "1.1.7.10" {
+		t.Fatalf("[DEPRECATED] bad node returned: %#v", out)
+	}
+	if out.CreateIndex != 1 || out.ModifyIndex != 7 {
+		t.Fatalf("[DEPRECATED] bad CreateIndex/ModifyIndex returned: %#v", out)
+	}
+}
```

Comment:
There is no test for the reverse case which is behaviour our test suite (and possibly other clients) rely on:

 - A node is registered _with_ an ID (by a real agent starting)
 - Later tooling re-registers the same node (say to update it's node meta) but doesn't bother to specify the `ID` field (because it never had to before)

The above should continue to work and replace the existing node with the new one _without an id_.

All that said, CI shows that the same tests are not broken _right now_ but they were only reliably broken before due to another change we reverted. They could still be non-deterministically broken since timing was involved.

Specifically in tests like `TestPrepaparedQuery_Execute`:
 - The test agent starts up assigns itself a UUID and _in the background_ Anti-entroypy _eventually_ registers that with the server (normally in milliseconds but that is enough of a delay to make it non-deterministic...)
 - Then actual test code like this: https://github.com/hashicorp/consul/blob/fa0d8aff335a5ea169b05cd2436358d2a77a07dc/agent/consul/prepared_query_endpoint_test.go#L1992
 - Re-registers the same node name to change it's health status. This registration happens not to have an ID so any of the following orderings can happen:
    - Agent AE gets there first so the _existing_ node in state _does_ have an ID while the incoming registration _doesn't_  - this is the case that is missing an explicit test here I think
    - test code registers first with no ID and then the AE upgrades it to have an ID. Then subsequent code in `setHealth` registers again without an ID.

It's pretty clear that the test is racey anyway regardless of the ID thing, but this is why we think it's important that we don't break the case where and existing node DOES have an ID but a registration doesn't specify one. We can always fix this test and arguably should, but other people might have legitimately built something that does something similar to this test and we can't break them.

Is that clear?

Now it _might_ be that your code is correct already and will always pass that test regardless of timing/interleaving - esp. given that it's not failing now. but I don't see an explicit test case here that verifies this behaviour and the logic is complicated enough I'm not sure I trust my self to reason about if it works alone 😄.

Could we add that?